File: javascript/file-to-text/webhook.js
import Replicate from "replicate";
import { EventEmitter } from "node:events";
import { Hono } from "hono";
import { serve } from "@hono/node-server";
import { readFile } from "node:fs/promises";
import { hostname } from "node:os";

// NOTE: Devbox must be public/unlisted for webhook to work.
const port = 3000;
const root = `https://${hostname}-${port}.csb.app`;
const callbackURL = `${root}/webhooks/replicate`;

// Use an EventEmitter to share the webhooks across the application.
const webhooks = new EventEmitter();

// Create our webserver for handling the webhooks.
// We use Hono, a web standards based framework here but this pattern
// applies to any web server like express, koa, hapi etc.
const app = new Hono();

// Replicate will POST a prediction to this endpoint.
app.post("/webhooks/replicate", async (c) => {
  // Extract the prediction from the request body.
  const prediction = await c.req.json();

  // Publish the prediction on the event bus, keyed by id.
  webhooks.emit(prediction.id, prediction);

  // Tell Replicate we received the request.
  c.status(200);
  return c.body("OK");
});

// Start the server.
const server = serve({ port, fetch: app.fetch });

// Setup Replicate.
const replicate = new Replicate();

const input = {
  image: await readFile("./image.png"),
  top_p: 1,
  prompt: "Are you allowed to swim here?",
  max_tokens: 1024,
  temperature: 0.2,
};
console.log("input image: %s, prompt: %s", "./image.png", input.prompt);

// Now we run the model and provide the callback url, filtering on the "completed" event.
console.log("Running...");
const prediction = await replicate.predictions.create({
  model: "yorickvp/llava-13b",
  version: "a0fdc44e4f2e1f20f2bb4e27846899953ac8e66c5886c5878fa1d6b73ce009e5",
  input,
  webhook: callbackURL,
  webhook_events_filter: ["completed"],
});

// Finally we wait for the webhooks EventEmitter to emit the completed prediction.
const completed = await new Promise((resolve) =>
  webhooks.once(prediction.id, resolve),
);
console.log(completed.output.join(""));

server.close();

File: python/file-to-text/webhook.py
import replicate
from aiohttp import web
import socket
import asyncio


async def main():
    # Create a Queue to handle the webhook events.
    webhook_queue = asyncio.Queue(1)

    # NOTE: The sandbox must be unlisted/public for the callback to work without auth
    port = 8000
    root = f"https://{socket.gethostname()}-{port}.csb.app"
    callback = f"{root}/callback/replicate"

    # Create a python webserver using aiohttp to handle the webhook and push
    # the completed prediction into our queue.
    routes = web.RouteTableDef()

    @routes.post('/callback/replicate')
    async def callback_replicate(request):
        prediction = await request.json()
        await webhook_queue.put(prediction)
        return web.Response(text="OK")

    app = web.Application()
    app.add_routes(routes)

    # Start the webserver.
    runner = web.AppRunner(app)
    await runner.setup()
    site = web.TCPSite(runner, port=port)
    await site.start()

    image = "./image.png"
    input = {
        # NOTE: We must pass 'rb' to read the file as binary.
        "image": open(image, 'rb'),
        "top_p": 1,
        "prompt": "Are you allowed to swim here?",
        "max_tokens": 1024,
        "temperature": 0.2,
    }
    print(f"input image: {image}, prompt: {input['prompt']}")

    print("Running...")
    await replicate.predictions.async_create(
        version="a0fdc44e4f2e1f20f2bb4e27846899953ac8e66c5886c5878fa1d6b73ce009e5",
        input=input,
        webhook=callback,
        webhook_events_filter=["completed"]
    )

    # Wait for the webhook event to complete and output the prediction.
    prediction = await webhook_queue.get()
    print("".join(prediction.get('output')))

    await site.stop()

asyncio.run(main())
